lib/vector-core/src/metrics/recency.rs

```diff
@@ -0,0 +1,367 @@
+    /// has not been updated recently enough, the key will be removed from the given registry if the
+    /// given generation also matches.
+    pub(super) fn should_store_gauge<S>(
+        &self,
+        key: &K,
+        gauge: &Generational<Arc<AtomicF64>>,
+        registry: &Registry<K, S>,
+    ) -> bool
+    where
+        S: Storage<K>,
+    {
+        self.should_store(
+            key,
+            gauge,
+            registry,
+            MetricKind::Gauge,
+            Registry::delete_gauge,
+        )
+    }
+
+    /// Checks if the given histogram should be stored, based on its known recency.
+    ///
+    /// If the given key has been updated recently enough, and should continue to be stored, this
+    /// method will return `true` and will update the last update time internally.  If the given key
+    /// has not been updated recently enough, the key will be removed from the given registry if the
+    /// given generation also matches.
+    pub(super) fn should_store_histogram<S>(
+        &self,
+        key: &K,
+        hist: &Generational<Arc<Histogram>>,
+        registry: &Registry<K, S>,
+    ) -> bool
+    where
+        S: Storage<K>,
+    {
+        self.should_store(
+            key,
+            hist,
+            registry,
+            MetricKind::Histogram,
+            Registry::delete_histogram,
+        )
+    }
+
+    fn should_store<F, S, T>(
+        &self,
+        key: &K,
+        value: &Generational<Arc<T>>,
+        registry: &Registry<K, S>,
+        kind: MetricKind,
+        delete_op: F,
+    ) -> bool
+    where
+        F: Fn(&Registry<K, S>, &K) -> bool,
+        S: Storage<K>,
+    {
+        let gen = value.get_generation();
+        if let Some(idle_timeout) = self.idle_timeout {
+            if self.mask.matches(kind) {
+                let mut guard = self.inner.lock();
+                let (clock, entries) = &mut *guard;
+
+                let now = clock.now();
+                let deleted = if let Some((last_gen, last_update)) = entries.get_mut(key) {
+                    // If the value is the same as the latest value we have internally, and
+                    // we're over the idle timeout period, then remove it and continue.
+                    if *last_gen == gen {
+                        // We don't want to delete the metric if there is an outstanding handle that
+                        // could later update the shared value. So, here we look up the count of
+                        // references to the inner value to see if there are more than expected.
+                        //
+                        // The magic value below comes from:
+                        // 1. The handle in the registry
+                        // 2. The handle held by the value
+                        // If there is another reference, then there is handle elsewhere.
+                        let referenced = Arc::strong_count(&value.inner) > 2;
```

Comment:
Hmmm, _greater_ than two?

The `Arc<...>`i held by the registry, the `Arc<...>` held by the handle itself... what's the third one?

Reply:
That's an unfortunate choice of words. The phrase is "the handle held by the value", where "value" refers to the variable `value` there, which is returned by `registry.get_X_handles`. I'll clarify.

Reply:
Ah, gotcha, right. 👍🏻